patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch

```diff
@@ -0,0 +1,695 @@
++
++using ::ABI::Windows::ApplicationModel::Core::ICoreApplication;
++using ::ABI::Windows::ApplicationModel::Core::ICoreApplicationView;
++using ::ABI::Windows::ApplicationModel::Core::ICoreImmersiveApplication;
++using ::ABI::Windows::Foundation::ActivateInstance;
++using ::ABI::Windows::Foundation::AsyncStatus;
++using ::ABI::Windows::Foundation::GetActivationFactory;
++using ::ABI::Windows::Foundation::IAsyncAction;
++using ::ABI::Windows::Foundation::IAsyncActionCompletedHandler;
++using ::ABI::Windows::Foundation::IAsyncInfo;
++using ::ABI::Windows::Media::Capture::IMediaCapture;
++using ::ABI::Windows::Media::Capture::IMediaCaptureFailedEventArgs;
++using ::ABI::Windows::Media::Capture::IMediaCaptureFailedEventHandler;
++using ::ABI::Windows::Media::Capture::IMediaCaptureInitializationSettings;
++using ::ABI::Windows::Media::Capture::IMediaCaptureInitializationSettings5;
++using ::ABI::Windows::Media::Capture::MediaCaptureMemoryPreference;
++using ::ABI::Windows::Media::Capture::StreamingCaptureMode;
++using ::ABI::Windows::Media::MediaProperties::IMediaRatio;
++using ::ABI::Windows::UI::Core::CoreDispatcherPriority;
++using ::ABI::Windows::UI::Core::CoreProcessEventsOption;
++using ::ABI::Windows::UI::Core::ICoreDispatcher;
++using ::ABI::Windows::UI::Core::ICoreWindow;
++using ::ABI::Windows::UI::Core::IDispatchedHandler;
++using ::Microsoft::WRL::AgileRef;
++using ::Microsoft::WRL::Callback;
++using ::Microsoft::WRL::ComPtr;
++using ::Microsoft::WRL::Delegate;
++using ::Microsoft::WRL::FtmBase;
++using ::Microsoft::WRL::Implements;
++using ::Microsoft::WRL::RuntimeClassFlags;
++using ::Microsoft::WRL::Wrappers::Event;
++using ::Microsoft::WRL::Wrappers::HString;
++using ::Microsoft::WRL::Wrappers::HStringReference;
+ 
+ namespace webrtc {
+ namespace videocapturemodule {
+ 
++HRESULT GetMediaCaptureForDevice(HSTRING device_id,
++                                  IMediaCapture** ppMediaCapture) {
++  HRESULT hr;
++
++  ComPtr<ICoreDispatcher> main_view_dispatcher;
++  ComPtr<IAsyncAction> async_action_ui_dispatcher;
++  ComPtr<IAsyncAction> async_action_media_capture;
++  ComPtr<IMediaCapture> media_capture;
++  AgileRef media_capture_agile;
++  AgileRef async_action_media_capture_agile;
++  Event event_wait_for_media_capture_async_action;
++  Event event_wait_for_media_capture_async_action_completion;
++  boolean has_thread_access;
++
++  // Acquires the main view dispacther (UI thread).
++  THR(GetMainViewDispatcher(&main_view_dispatcher));
++
++  // We'll be dispatching and waiting for code from the UI thread.
++  // Let's make sure this is not the UI thread.
++  THR(main_view_dispatcher->get_HasThreadAccess(&has_thread_access));
++  THR(has_thread_access ? E_FAIL : S_OK);
++
++  // The media capture device has to be initialized in the UI thread.
++  // The following event is used for letting this thread know that
++  // async_action_media_capture has been created.
++  event_wait_for_media_capture_async_action =
```

Comment:
Use `rtc::Event`.

Reply:
-2 lines of code, consistency of codebase, and familiarity for non-Windows WebRTC devs. I personally have no idea what all those options in `CreateEventEx()` are; to review I need to pop up the documentation.